src/x.c

```diff
@@ -93,21 +94,77 @@ xcb_window_t wid_get_prop_window(session_t *ps, xcb_window_t wid, xcb_atom_t apr
  */
 bool wid_get_text_prop(session_t *ps, xcb_window_t wid, xcb_atom_t prop, char ***pstrlst,
                        int *pnstr) {
-	XTextProperty text_prop = {NULL, XCB_NONE, 0, 0};
+	xcb_generic_error_t *e = NULL;
+	auto r = xcb_get_property_reply(
+	    ps->c, xcb_get_property(ps->c, 0, wid, prop, XCB_ATOM_ANY, 0, 0), &e);
+	if (!r) {
+		log_debug_x_error(e, "Failed to get window property for %#010x", wid);
+		free(e);
+		return false;
+	}
+
+	auto type = r->type;
+	auto format = r->format;
+	auto length = r->bytes_after;
+	free(r);
+
+	if (type == XCB_ATOM_NONE) {
+		return false;
+	}
+
+	if (type != XCB_ATOM_STRING && type != ps->atoms->aUTF8_STRING &&
+	    type != ps->atoms->aC_STRING) {
+		log_warn("Text property %d of window %#010x has unsupported type: %d",
+		         prop, wid, type);
+		return false;
+	}
 
-	if (!(XGetTextProperty(ps->dpy, wid, &text_prop, prop) && text_prop.value))
+	if (format != 8) {
+		log_warn("Text property %d of window %#010x has unexpected format: %d",
+		         prop, wid, format);
 		return false;
+	}
 
-	if (Success != XmbTextPropertyToTextList(ps->dpy, &text_prop, pstrlst, pnstr) ||
-	    !*pnstr) {
-		*pnstr = 0;
-		if (*pstrlst)
-			XFreeStringList(*pstrlst);
-		XFree(text_prop.value);
+	r = xcb_get_property_reply(
+	    ps->c, xcb_get_property(ps->c, 0, wid, prop, type, 0, length), &e);
+	if (!r) {
+		log_debug_x_error(e, "Failed to get window property for %#010x", wid);
+		free(e);
 		return false;
 	}
 
-	XFree(text_prop.value);
+	assert(length == (uint32_t)xcb_get_property_value_length(r));
+
+	void *data = xcb_get_property_value(r);
+	unsigned int nstr = 0;
+	uint32_t current_offset = 0;
+	while (current_offset < length) {
+		current_offset +=
+		    (uint32_t)strnlen(data + current_offset, length - current_offset) + 1;
+		nstr += 1;
+	}
+
+	// Allocate the pointers and the strings together
+	void *buf = NULL;
+	if (posix_memalign(&buf, alignof(void *), length + sizeof(void *) * nstr + 1) != 0) {
+		abort();
+	}
+
+	char *strlst = buf + sizeof(void *) * nstr;
+	memcpy(strlst, xcb_get_property_value(r), length);
+	strlst[length] = '\0';        // X strings aren't guaranteed to be null terminated
+
+	char **ret = buf;
+	current_offset = 0;
+	nstr = 0;
+	while (current_offset < length) {
+		ret[nstr] = strlst + current_offset;
```

Comment:
We implicitly assume `sizeof(void *) == sizeof(char *)`. This should be fine with most systems but isn't required by the C standard and could differ.
Since we intend to return an array of `char *` we could allocate it as as such.

Reply:
Addressed